### PR TITLE
RFD #4: Plugin API

### LINKED ISSUE #11: Write the "Plugin Architecture API" RFD
Once we've characterized the expected end state for Hipcheck's new plugin architecture, we ought to then specify what the API would look like. This involves multiple "parts" of the API, which I will decompose here, taking [inspiration from the design of `dtrace`](https://opensource.apple.com/source/xnu/xnu-3789.41.3/bsd/sys/dtrace.h.auto.html) (the dynamic tracing tool originally designed for Solaris at Sun, now ported to a number of other kernels and operating systems). The `dtrace` implementation breaks down the following parts of the API:

* `dtrace`-to-Provider ("Provider" is what `dtrace` calls their plugins, which enable instrumentation of the system)
    * This is the API that `dtrace` exposes to Providers
* Provider-to-`dtrace`
    * This is how Providers register themselves with `dtrace` and tell `dtrace` how to use them

To analogize to Hipcheck, this would basically become:

* Hipcheck-to-Plugin
    * When a plugin is registering itself, the plugin would provide a structure of functions to Hipcheck which describe the operations necessary for performing the operations provided by the plugin.
* Plugin-to-Hipcheck
    * The API Hipcheck would expose to plugins to enable registration of new plugins with Hipcheck

`dtrace` also has a mechanism for "meta providers" which is how `dtrace` supports Userland Statically-Defined Tracing (USDT), for which there wouldn't be an analogue in our context.

To put the above breakdown another way, there needs to be:

* A way for plugins to be registered, so Hipcheck knows about them and can call them.
* A way for plugins to enable Hipcheck to call the functions the plugins provide.

Note that this RFD should still be agnostic about the specific way that plugins are implemented. We'll want to explore multiple options for implementing plugins, including:

* WebAssembly
* Separate binaries in their own subprocess with inter-process communication between them and Hipcheck
* C FFI with dynamic linking

Exploration of these options will be addressed by other RFD's.

### PR DESCRIPTION
**Pull request by alilleybrinker**
Closes #11.

Edit: **[__Rendered__](https://github.com/mitre/hipcheck/blob/416530d6e496d7c9a0ef756c7ec32292063bd529/site/content/rfds/0004-plugin-api.md)**

### CODE REVIEW

File: site/content/rfds/0004-plugin-api.md [hunk 1]
```diff
@@ -0,0 +1,932 @@
...
+```kdl
+publisher "mitre"
+name "affiliation"
+version "0.1.0"
+license "Apache-2.0"
+entrypoint {
+  on arch="aarch64-apple-darwin" "./hc-mitre-affiliation"
+  on arch="x86_64-apple-darwin" "./hc-mitre-affiliation"
+  on arch="x86_64-unknown-linux-gnu" "./hc-mitre-affiliation"
+  on arch="x86_64-pc-windows-msvc" "./hc-mitre-affiliation"
```
Comment (mchernicoff) [inline site/content/rfds/0004-plugin-api.md:119]: Just confirming that that would be the correct entrypoint even on Windows.

File: site/content/rfds/0004-plugin-api.md [hunk 2]
```diff
@@ -0,0 +1,932 @@
...
+
+The RFD so far has described how plugins are defined by plugin creators, and
+the API expected of plugins in order for Hipcheck to interact with them. What
+has _not_ been described is how end-users of Hipcheck should use plugins
+themselves. This section answers that question.
+
+### Policy Files
+
+Today, Hipcheck determines what analyses to run and how to configure those
+analyses a "configuration file," and when you first install Hipcheck you
```
Comment (tpepper-gov) [inline site/content/rfds/0004-plugin-api.md:393]: sentence fragment missing word(s?)

File: site/content/rfds/0004-plugin-api.md [hunk 3]
```diff
@@ -0,0 +1,932 @@
...
+
+The __plugin discovery manifest__ is an additional manifest which tells
+Hipcheck where to find the correct plugin artifact for a specific target
+architecture. In the future, Hipcheck's plugin system may support one or
+more plugin registries which ease discovery of Hipcheck plugins. In the
+short term, the standard mechanism for discovery of Hipcheck plugins
+will be for plugin producers to distribute a per-plugin discovery manifest
+which contains version and target architecture information necessary for
+Hipcheck to download the _correct_ plugin executable artifact to run
+on the end-user's host.
```
Comment (j-lanson) [inline site/content/rfds/0004-plugin-api.md:135]: Do we want to consider at this time the possibility of some users wanting to run in a fully-offline mode with plugins they already have on-disk?

File: site/content/rfds/0004-plugin-api.md (hunk 1 above)
Comment (alilleybrinker) [reply site/content/rfds/0004-plugin-api.md:119]: Thanks! I do think that we'll likely want to enhance the target triple stuff to permit some "grouping expressions" when granularly listing every target triple is too tedious. Rust has a version of this in the `cfg` system where you can just say "windows" for example and cover _all_ windows targets.

File: site/content/rfds/0004-plugin-api.md (hunk 2 above)
Comment (alilleybrinker) [reply site/content/rfds/0004-plugin-api.md:393]: Ah good catch!

File: site/content/rfds/0004-plugin-api.md (hunk 3 above)
Comment (alilleybrinker) [reply site/content/rfds/0004-plugin-api.md:135]: Good point! I think this would properly go in the "Running Plugins" section.

**Review by mchernicoff — APPROVED** [review]

**Pull request merged**